### PR TITLE
Add fixed-wing dynamics plugin

### PR DESCRIPTION
**Pull request by pvechersky**
Adding a plugin for fixed-wing aerodynamics that can receive control inputs from a joystick using the 'rotors_joy_interface' or the RC using 'rotors_hil_interface'. Tested with a USB joystick using the new 'fixed_wing_with_joy.launch' file to ensure the model (at least the Techpod model and the provided parameters) flies.

Some notes:
- To avoid reliance on ROS, the parameters are parsed with yaml-cpp instead of being loaded with rosparam.
- While we do have a wind plugin that publishes wind speed directly on a Gazebo topic, I thought it might still be good to have the fixed-wing dynamics plugin reach out to the ROS topic (via ros interface plugin) in case wind speed comes from some external program.
- At the moment it is a bit difficult to keep track of the model while flying it in such a big world, and Gazebo's 'follow' feature is not always adequate, but I'm hoping to submit a view control plugin soon.
- Since the control surface joints are not actuated yet (which would be strictly for visualization purposes) I set their joint limits to zeros for now, otherwise they 'flap' around a little bit during flight from inertia.

@tstastny @philipoe

### CODE REVIEW

File: rotors_gazebo/models/yosemite/yosemite.sdf [hunk 1]
```diff
@@ -8,7 +8,7 @@
           <heightmap>
             <uri>model://yosemite/materials/textures/yosemite.png</uri>
             <size>1000 1000 125</size>
-            <pos>100 0 -30</pos>
+            <pos>0 0 -15</pos>
```
Comment (hitimo) [inline rotors_gazebo/models/yosemite/yosemite.sdf]: Out of curiosity: What does this change?

File: rotors_gazebo/launch/fixed_wing_with_joy.launch [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  <arg name="enable_ground_truth" default="true"/>
+  <arg name="log_file" default="$(arg uav_name)"/>
+
+  <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models"/>
+  <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models"/>
+  <include file="$(find gazebo_ros)/launch/empty_world.launch">
+    <arg name="world_name" value="$(find rotors_gazebo)/worlds/$(arg world_name).world"/>
+    <!-- <arg name="debug" value="true" /> -->
+    <arg name="paused" value="true" />
+    <!-- <arg name="gui" value="false" /> -->
```
Comment (hitimo) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: rm? Or does it have any purpose?

File: rotors_gazebo/launch/fixed_wing_with_joy.launch [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  <arg name="world_name" default="yosemite"/>
+  <arg name="enable_logging" default="false"/>
+  <arg name="enable_ground_truth" default="true"/>
+  <arg name="log_file" default="$(arg uav_name)"/>
+
+  <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models"/>
+  <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models"/>
+  <include file="$(find gazebo_ros)/launch/empty_world.launch">
+    <arg name="world_name" value="$(find rotors_gazebo)/worlds/$(arg world_name).world"/>
+    <!-- <arg name="debug" value="true" /> -->
```
Comment (hitimo) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: rm? Or does it have any purpose?

File: rotors_gazebo/worlds/yosemite.world [hunk 4]
```diff
@@ -181,7 +186,7 @@
         </material>
       </visual>
       <visual name='visual_7'>
-        <pose>475 125 0 0 0 0</pose>
+        <pose>375 125 0 0 0 0</pose>
```
Comment (hitimo) [inline rotors_gazebo/worlds/yosemite.world]: Why these changes here and above?

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/fw_parameters.h [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+  Eigen::Vector2d c_pitch_moment_delta_elv;
+
+  Eigen::Vector2d c_yaw_moment_beta;
+  Eigen::Vector2d c_yaw_moment_r;
+  Eigen::Vector2d c_yaw_moment_delta_rud;
+
+  Eigen::Vector3d c_thrust;
+
+  void LoadAeroParamsYAML(const std::string& yaml_path) {
+    YAML::Node node = YAML::LoadFile(yaml_path);
```
Comment (hitimo) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/fw_parameters.h]: Can it be const?

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_fw_dynamics_plugin.h [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ROTORS_GAZEBO_PLUGINS_FW_DYNAMICS_PLUGIN_H
+#define ROTORS_GAZEBO_PLUGINS_FW_DYNAMICS_PLUGIN_H
+
+#include <gazebo/common/Plugin.hh>
+#include <gazebo/common/common.hh>
```
Comment (hitimo) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_fw_dynamics_plugin.h]: alphabetize

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_fw_dynamics_plugin.h [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    GzRollPitchYawrateThrustMsgPtr;
+typedef const boost::shared_ptr<const gz_mav_msgs::WindSpeed>
+    GzWindSpeedMsgPtr;
+
+// Default values.
+static constexpr bool kDefaultIsInputJoystick = false;
+
+// Constants.
+static constexpr double kAirDensity = 1.18;
+static constexpr double kG = 9.81;
```
Comment (hitimo) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_fw_dynamics_plugin.h]: I'd prefer kGravity

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+  // Get the robot namespace.
+  if (_sdf->HasElement("robotNamespace"))
+    namespace_ = _sdf->GetElement("robotNamespace")->Get<std::string>();
+  else
+    gzerr << "[gazebo_fw_dynamics_plugin] Please specify a robotNamespace.\n";
+
+  // Create the node handle.
+  node_handle_ = transport::NodePtr(new transport::Node());
+
+  // Initisalise with default namespace (typically /gazebo/default/).
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: typo

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+  Eigen::Matrix3d R_Wind_B;
+  R_Wind_B << ca * cb, sb, sa * cb,
+              -sb * ca, cb, -sa * sb,
+              -sa, 0.0, ca;
+
+  Eigen::Matrix3d R_Wind_B_t = R_Wind_B.transpose();
+
+  // Transform all the forces and moments into the body frame
+  Eigen::Vector3d forces_B = R_Wind_B_t * forces_Wind + force_thrust_B;
+  Eigen::Vector3d moments_B = R_Wind_B_t * moments_Wind;
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: const ref?
Many of the variables (`const double ...`) above could also be const.

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+    connect_ros_to_gazebo_topic_msg.set_msgtype(
+        gz_std_msgs::ConnectRosToGazeboTopic::ACTUATORS);
+  }
+
+  gz_connect_ros_to_gazebo_topic_pub->Publish(connect_ros_to_gazebo_topic_msg,
+                                              true);
+}
+
+void GazeboFwDynamicsPlugin::ActuatorsCallback(
+    GzActuatorsMsgPtr &actuators_msg) {
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: `GzActuatorsMsgPtr& actuators_msg`

Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: Can it be `const GzActuatorsMsgPtr& actuators_msg`?

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+  delta_flap_ = NormalizedInputToAngle(vehicle_params_.flap,
+      actuators_msg->normalized(vehicle_params_.flap.channel));
+  delta_rudder_ = NormalizedInputToAngle(vehicle_params_.rudder,
+      actuators_msg->normalized(vehicle_params_.rudder.channel));
+
+  throttle_ = actuators_msg->normalized(vehicle_params_.throttle_channel);
+}
+
+void GazeboFwDynamicsPlugin::RollPitchYawrateThrustCallback(
+    GzRollPitchYawrateThrustMsgPtr& roll_pitch_yawrate_thrust_msg) {
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: Can it be cref?

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+        roll_pitch_yawrate_thrust_sub_topic_);
+    connect_ros_to_gazebo_topic_msg.set_gazebo_topic("~/" + namespace_ + "/" +
+        roll_pitch_yawrate_thrust_sub_topic_);
+    connect_ros_to_gazebo_topic_msg.set_msgtype(
+        gz_std_msgs::ConnectRosToGazeboTopic::ROLL_PITCH_YAWRATE_THRUST);
+  } else {
+    // ============================================ //
+    // ===== ACTUATORS MSG SETUP (ROS->GAZEBO) ==== //
+    // ============================================ //
+
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: rm vertical space

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+  // Create temporary "ConnectRosToGazeboTopic" publisher and message
+  gazebo::transport::PublisherPtr gz_connect_ros_to_gazebo_topic_pub =
+      node_handle_->Advertise<gz_std_msgs::ConnectRosToGazeboTopic>(
+          "~/" + kConnectRosToGazeboSubtopic, 1);
+  gz_std_msgs::ConnectRosToGazeboTopic connect_ros_to_gazebo_topic_msg;
+
+  // ============================================ //
+  // ==== WIND SPEED MSG SETUP (ROS->GAZEBO) ==== //
+  // ============================================ //
+
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: rm vertical space

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 14]
```diff
@@ -318,6 +318,26 @@ void GazeboRosInterfacePlugin::GzConnectRosToGazeboTopicMsgCallback(
...
+              gz_connect_ros_to_gazebo_topic_msg->ros_topic(), 1,
+              boost::bind(
+                  &GazeboRosInterfacePlugin::
+                      RosRollPitchYawrateThrustMsgCallback,
+                  this, _1, gz_publisher_ptr));
+
+      // Save reference to the ROS subscriber so callback will continue to be
+      // called.
+      ros_subscribers.push_back(ros_subscriber);
+
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: rm vertical space

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 15]
```diff
@@ -786,7 +806,31 @@ void GazeboRosInterfacePlugin::GzTransformStampedMsgCallback(
...
+  // ============================================ //
+  // =================== HEADER ================= //
+  // ============================================ //
+  ConvertHeaderGzToRos(gz_twist_stamped_msg->header(),
+                       &ros_twist_stamped_msg_.header);
+
+  // ============================================ //
+  // =================== TWIST ================== //
+  // ============================================ //
+
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: rm vertical space

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+  // B denotes body frame and W world frame ... e.g., W_rot_W_B denotes
+  // rotation of B wrt. W expressed in W.
+  math::Quaternion W_rot_W_B = link_->GetWorldPose().rot;
+  math::Vector3 B_air_speed_W_B = W_rot_W_B.RotateVectorReverse(
+      link_->GetWorldLinearVel() - W_wind_speed_W_B_);
+  math::Vector3 B_angular_velocity_W_B = link_->GetRelativeAngularVel();
+
+  // Traditionally, fixed-wing aerodynamics use NED (North-East-Down) frame,
+  // but since our model's body frame is in North-West-Up frame we rotate the
+  // linear and angular velocities by 180 degrees around the X axis.
```
Comment (hitimo) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: Thanks for the comment :+1:

**Comment by hitimo** [comment]
```
/var/lib/jenkins/jobs/rotors_simulator/workspace/label/master/src/rotors_simulator/rotors_gazebo_plugins/include/rotors_gazebo_plugins/fw_parameters.h:181:23: error: ?LoadFile? is not a member of ?YAML?
```

File: rotors_gazebo/models/yosemite/yosemite.sdf (hunk 1 above)
Comment (pvechersky) [inline rotors_gazebo/models/yosemite/yosemite.sdf]: With previous coordinates, I noticed what seemed to be a discrepancy between the visual and collison layer of the heightmap - I would sometimes be flying through seemingly free space but the aircraft would crash/stop as if it hit something. Seems to work better with the new coordinates. Plus, as a general practice I think it's clearer when the link origin is aligned with the model origin if possible (at least in the xy-plane in this case).

File: rotors_gazebo/worlds/yosemite.world (hunk 4 above)
Comment (pvechersky) [inline rotors_gazebo/worlds/yosemite.world]: Related to the change to 'yosemite.sdf' commented on above.

File: rotors_gazebo/launch/fixed_wing_with_joy.launch (hunk 2 above)
Comment (pvechersky) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: These commented out arguments are present in pretty much all the RotorS launch files, and have been for a while, so I was being consistent with the rest of them. I think they're there to make it easier for users to know that such arguments do exist if needed. Or maybe a left over artifact from development. I think @ffurrer is better to ask :)

File: rotors_gazebo/launch/fixed_wing_with_joy.launch (hunk 3 above)
Comment (pvechersky) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: Same as above.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_fw_dynamics_plugin.h (hunk 6 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_fw_dynamics_plugin.h]: I spotted this recent commit: #351 and followed the same order ... If the change in this commit was intentional it would suggest the capitalized files are grouped together vs non-capitilized. However, this is not consistent across all the current plugins, so hard to say.

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp (hunk 8 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: Oops. This is why they say not to copy-paste code :) ... I'll fix this in other plugins, too.

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp (hunk 10 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: 'GzActuatorsMsgPtr' is already typedef'ed as 'const boost::shared_ptr<const gz_sensor_msgs::Actuators>' in the header for shorter notation, much like it was recently done in other plugins.

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp (hunk 11 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: Same as above.

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp (hunk 12 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: For many things, especially those involving the Gazebo-ROS interfacing, I followed the formatting of other plugins. So perhaps it's best to consult @gbmhunter regarding this.

File: rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp (hunk 13 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_fw_dynamics_plugin.cpp]: Same as above.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 14 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Same as above.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 15 above)
Comment (pvechersky) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Same as above.

**Comment by pvechersky** [comment]
Is the yaml-cpp package installed on the build server?

**Comment by hitimo** [comment]
test this please

**Comment by hitimo** [comment]
test this please

**Comment by tstastny** [comment]
@pvechersky FW dynamics look good to me! Thanks for adding the capability to add a user specified yaml for anyone who wants to put in their own coefficients. The comments are also good.

File: rotors_gazebo/models/yosemite/yosemite.sdf (hunk 1 above)
Comment (hitimo) [inline rotors_gazebo/models/yosemite/yosemite.sdf]: Great, thanks.

File: rotors_gazebo/launch/fixed_wing_with_joy.launch (hunk 2 above)
Comment (hitimo) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: Ok for me.

**Comment by hitimo** [comment]
@pvechersky I added https://github.com/ethz-asl/yaml_cpp_catkin as dependency for Jenkins. But since `yaml_cpp_catkin` is not listed as dependency in package.xml, it's not pulled in and fails. I'd propose to add `yaml_cpp_catkin` to package.xml.

File: rotors_gazebo/launch/fixed_wing_with_joy.launch (hunk 2 above)
Comment (ffurrer) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: Yes, this was left there, as you thought, for one to know how to start gdb and how to turn the GUI off. Both of them I used frequently. But if no one else uses them, or they are obvious, I'm also open to remove them.

Comment (pvechersky) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: I also make use these quite often during development, so I'd be all for leaving them there.

**Comment by pvechersky** [comment]
@hitimo Thank you for looking into this and adding the dependency for Jenkins! This would work for me.

If everybody else is OK with it then I think in addition to the package.xml and CMakeLists.txt I would also need to update the .rosinstall files to pull the repo and also the README with the instructions, right?

**Comment by ffurrer** [comment]
@pvechersky yup, I think there is no issue with adding the `yaml_cpp_catkin` dependency, or @burrimi do you have an opinion on  this?
If @burrimi is fine with it, updating the README and the .rosinstall would then of course be great! Thanks a lot :)

File: rotors_gazebo/launch/fixed_wing_with_joy.launch (hunk 2 above)
Comment (hitimo) [inline rotors_gazebo/launch/fixed_wing_with_joy.launch]: Sure, let's leave it.

**Comment by pvechersky** [comment]
@hitimo @ffurrer I added yaml_cpp_catkin as a dependency and included it in the README and .rosinstall files. I also had to include catkin_simple in that as well since yaml_cpp_catkin depends on it.
